File: frontend/websites/puconvocation-com/src/controllers/AttendeeController.ts
/*
 * Copyright (C) PU Convocation Management System Authors
 *
 * This software is owned by PU Convocation Management System Authors.
 * No part of the software is allowed to be copied or distributed
 * in any form. Any attempt to do so will be considered a violation
 * of copyright law.
 *
 * This software is protected by copyright law and international
 * treaties. Unauthorized copying or distribution of this software
 * is a violation of these laws and could result in severe penalties.
 */

import {
  Attendee,
  AttendeesInEnclosure,
  AttendeeWithEnclosureMetadata,
  AttendeeWithPagination,
  Response,
} from "@dto/index";
import { HttpService } from "@services/index";

export default class AttendeeController {
  private BASE_URL = process.env.NEXT_PUBLIC_DYNAMICS_SERVICE_URL as string;

  private httpService: HttpService;

  private ATTENDEE_ROUTES = this.BASE_URL.concat("/attendees");

  public constructor(options?: { cookies?: string }) {
    this.httpService = new HttpService(this.BASE_URL, options);
  }

  public async getAttendee(
    identifier: string,
  ): Promise<Response<AttendeeWithEnclosureMetadata, string>> {
    return await this.httpService.get<AttendeeWithEnclosureMetadata>(
      `${this.ATTENDEE_ROUTES}/${identifier}`,
    );
  }

  public async getTotalAttendeesCount(): Promise<
    Response<{ count: number }, string>
  > {
    return await this.httpService.get<{ count: number }>(
      `${this.ATTENDEE_ROUTES}/totalCount`,
    );
  }

  public async getAllAttendees(
    page: number,
    limit: number,
  ): Promise<Response<AttendeeWithPagination, string>> {
    const queryParams = new URLSearchParams({
      page: page.toString(),
      limit: limit.toString(),
    });

    return await this.httpService.get<AttendeeWithPagination>(
      `${this.ATTENDEE_ROUTES}/all?${queryParams.toString()}`,
    );
  }

  public async searchAttendees(
    query: string,
  ): Promise<Response<Array<Attendee>, string>> {
    const queryParams = new URLSearchParams({
      query: query.trim(),
    });

    return await this.httpService.get<Array<Attendee>>(
      `${this.ATTENDEE_ROUTES}/search?${queryParams.toString()}`,
    );
  }

  public async getAttendeeFromVerificationToken(
    token: string,
  ): Promise<Response<Attendee, string>> {
    return await this.httpService.get<Attendee>(
      `${this.ATTENDEE_ROUTES}/verificationToken/${token}`,
    );
  }

  public async uploadAttendeeList(
    file: File,
  ): Promise<Response<string, string>> {
    const form = new FormData();
    form.append(file.name, file);
    return await this.httpService.post<string>(
      `${this.ATTENDEE_ROUTES}/upload`,
      form,
    );
  }

  public async createTransaction(
    studentEnrollmentNumber: string,
  ): Promise<Response<AttendeeWithEnclosureMetadata, string>> {
    return await this.httpService.post<AttendeeWithEnclosureMetadata>(
      `${this.BASE_URL}/transactions/new`,
      {
        studentEnrollmentNumber: studentEnrollmentNumber,
      },
      {
        expectedStatusCode: 200,
      },
    );
  }

  public async mutateAttendeeLock(
    locked: boolean,
  ): Promise<Response<any, string>> {
    return await this.httpService.post<any>(
      `${this.ATTENDEE_ROUTES}/mutateAttendeeLock?locked=${locked}`,
    );
  }

  public async attendeesInEnclosure(
    enclosure: string,
  ): Promise<Response<AttendeesInEnclosure, string>> {
    return await this.httpService.get<any>(
      `${this.ATTENDEE_ROUTES}/enclosure/${enclosure}`,
    );
  }
}

File: backend/services/dynamics-service/src/main/kotlin/com/puconvocation/controllers/AttendeeController.kt
/*
 * Copyright (C) PU Convocation Management System Authors
 *
 * This software is owned by PU Convocation Management System Authors.
 * No part of the software is allowed to be copied or distributed
 * in any form. Any attempt to do so will be considered a violation
 * of copyright law.
 *
 * This software is protected by copyright law and international
 * treaties. Unauthorized copying or distribution of this software
 * is a violation of these laws and could result in severe penalties.
 */

package com.puconvocation.controllers

import com.puconvocation.commons.dto.AttendeeWithEnclosureMetadata
import com.puconvocation.commons.dto.ErrorResponse
import com.puconvocation.database.mongodb.entities.Attendee
import com.puconvocation.database.mongodb.entities.RemoteConfig
import com.puconvocation.database.mongodb.repositories.AttendeeRepository
import com.puconvocation.database.mongodb.repositories.RemoteConfigRepository
import com.puconvocation.enums.ResponseCode
import com.puconvocation.serializers.CSVSerializer
import com.puconvocation.services.AuthService
import com.puconvocation.services.DistributedLock
import com.puconvocation.services.LambdaService
import io.ktor.http.*
import io.ktor.http.content.*
import io.ktor.utils.io.*
import io.ktor.utils.io.core.*
import org.apache.commons.io.FilenameUtils
import java.io.ByteArrayInputStream
import java.io.InputStreamReader
import java.time.Duration
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit

class AttendeeController(
    private val attendeeRepository: AttendeeRepository,
    private val csvSerializer: CSVSerializer,
    private val authService: AuthService,
    private val distributedLock: DistributedLock,
    private val lambdaService: LambdaService,
    private val remoteConfigRepository: RemoteConfigRepository,
) {
    suspend fun getAttendee(identifier: String): Result<AttendeeWithEnclosureMetadata, ErrorResponse> {
        if (!remoteConfigRepository.getConfig().attendees.locked) {
            return Result.Error(
                httpStatusCode = HttpStatusCode.NotFound,
                error = ErrorResponse(
                    errorCode = ResponseCode.ATTENDEE_NOT_FOUND,
                    message = "Could not find attendee for identifier $identifier"
                )

            )
        }

        val attendee = attendeeRepository.getAttendeeWithEnclosureMetadata(identifier)
            ?: return Result.Error(
                httpStatusCode = HttpStatusCode.NotFound,
                error = ErrorResponse(
                    errorCode = ResponseCode.ATTENDEE_NOT_FOUND,
                    message = "Could not find attendee for identifier $identifier"
                )

            )

        return Result.Success(
            attendee
        )
    }

    suspend fun uploadAttendees(
        authorizationToken: String?,
        multiPart: MultiPartData
    ): Result<HashMap<String, Any>, ErrorResponse> {

        if (!authService.isAuthorized(
                role = "write:Attendee",
                principal = authorizationToken
            )
        ) {
            return Result.Error(
                httpStatusCode = HttpStatusCode.Forbidden,
                error = ErrorResponse(
                    errorCode = ResponseCode.NOT_PERMITTED,
                    message = "You don't have privilege to upload attendee details."
                )

            )
        }

        val lockAcquired = distributedLock.acquire("attendeeUploadLock", Duration.of(5, ChronoUnit.MINUTES))

        if (!lockAcquired) {
            return Result.Error(
                httpStatusCode = HttpStatusCode.BadRequest,
                error = ErrorResponse(
                    errorCode = ResponseCode.REQUEST_NOT_FULFILLED,
                    message = "Couldn't upload attendees list."
                )
            )
        }

        if (remoteConfigRepository.getConfig().attendees.locked) {
            distributedLock.release("attendeeUploadLock")
            return Result.Error(
                httpStatusCode = HttpStatusCode.BadRequest,
                error = ErrorResponse(
                    errorCode = ResponseCode.ALREADY_LOCKED,
                    message = "Couldn't upload attendees list. The Attendees List is already locked."
                )
            )
        }

        val part = multiPart.readPart()
        if (part !is PartData.FileItem ||
            !FilenameUtils.getExtension(part.originalFileName).equals("csv", ignoreCase = true)
        ) {
            distributedLock.release("attendeeUploadLock")
            return Result.Error(
                httpStatusCode = HttpStatusCode.BadRequest,
                error = ErrorResponse(
                    errorCode = ResponseCode.INVALID_FILE_FORMAT,
                    message = "Invalid file format. The file must of type CSV"
                )
            )
        }

        val attendees: List<Attendee> = csvSerializer.serializeAttendeesCSV(
            InputStreamReader(
                ByteArrayInputStream(
                    part.provider.invoke().readBuffer().readBytes()
                ),
                "UTF-8"
            )
        )
        val response = attendeeRepository.uploadAttendees(attendees)
        if (!response) {
            distributedLock.release("attendeeUploadLock")
            return Result.Error(
                httpStatusCode = HttpStatusCode.NotModified,
                error = ErrorResponse(
                    errorCode = ResponseCode.FILE_NOT_UPLOADED,
                    message = "Could not upload attendees at the moment."
                )

            )
        }


        lambdaService.invoke("PUConvocationAttendeeJob")
        distributedLock.release("attendeeUploadLock")

        return Result.Success(
            hashMapOf(
                "code" to ResponseCode.FILE_UPLOADED,
                "message" to "File uploaded successfully"
            )
        )
    }

    suspend fun getTotalAttendees(): Result<HashMap<String, Any>, ErrorResponse> {
        val totalCount = attendeeRepository.getTotalAttendees()
        return Result.Success(
            hashMapOf("count" to totalCount)
        )
    }

    suspend fun getAttendeeFromVerificationToken(
        authorizationToken: String?,
        token: String
    ): Result<Attendee, ErrorResponse> {
        if (!authService.isAuthorized(
                role = "read:Attendee",
                principal = authorizationToken
            )
        ) {
            return Result.Error(
                httpStatusCode = HttpStatusCode.Forbidden,
                error = ErrorResponse(
                    errorCode = ResponseCode.NOT_PERMITTED,
                    message = "You don't have privilege to Lock attendee details."
                )

            )
        }

        val attendee = attendeeRepository.getAttendeeFromVerificationToken(token = token) ?: return Result.Error(
            httpStatusCode = HttpStatusCode.NotFound,
            error = ErrorResponse(
                errorCode = ResponseCode.ATTENDEE_NOT_FOUND,
                message = "Could not find attendee"
            )

        )

        return Result.Success(
            attendee
        )
    }


    suspend fun mutateAttendeeLock(
        authorizationToken: String?,
        lock: Boolean
    ): Result<HashMap<String, Any>, ErrorResponse> {
        if (!authService.isAuthorized(
                role = "write:Attendee",
                principal = authorizationToken
            )
        ) {
            return Result.Error(
                httpStatusCode = HttpStatusCode.Forbidden,
                error = ErrorResponse(
                    errorCode = ResponseCode.NOT_PERMITTED,
                    message = "You don't have privilege change attendee lock state."
                )
            )
        }

        val lockAcquired = distributedLock.acquire("attendeeLock", Duration.of(5, ChronoUnit.MINUTES))

        if (!lockAcquired) {
            return Result.Error(
                httpStatusCode = HttpStatusCode.BadRequest,
                error = ErrorResponse(
                    errorCode = ResponseCode.ALREADY_LOCKED,
                    message = "The Attendees List couldn't be ${if (lock) "locked" else "unlocked"}. Please try again later."
                )
            )
        }

        if (remoteConfigRepository.getConfig().attendees.locked && ChronoUnit.DAYS.between(
                remoteConfigRepository.getConfig().attendees.updatedAt,
                LocalDateTime.now()
            ) <= 3
        ) {
            distributedLock.release("attendeeLock")
            return Result.Error(
                httpStatusCode = HttpStatusCode.BadRequest,
                error = ErrorResponse(
                    errorCode = ResponseCode.NOT_PERMITTED,
                    message = "Attendees List cannot be unlocked before 3 days of locking it."
                )
            )
        }

        if (lock == remoteConfigRepository.getConfig().attendees.locked) {
            distributedLock.release("attendeeLock")
            return Result.Error(
                httpStatusCode = HttpStatusCode.BadRequest,
                error = ErrorResponse(
                    errorCode = if (lock) ResponseCode.ALREADY_LOCKED else ResponseCode.ALREADY_UNLOCKED,
                    message = "Attendees List is already ${if (lock) "locked" else "unlocked"}."
                )
            )
        }

        val acknowledge = remoteConfigRepository.mutateAttendeeLock(
            RemoteConfig.Attendees(
                locked = lock,
                updatedAt = LocalDateTime.now()
            )
        )

        if (!acknowledge) {
            distributedLock.release("attendeeLock")
            return Result.Error(
                httpStatusCode = HttpStatusCode.BadRequest,
                error = ErrorResponse(
                    errorCode = ResponseCode.REQUEST_NOT_FULFILLED,
                    message = "The Attendees List couldn't be ${if (lock) "locked" else "unlocked"}. Please try again later."
                )
            )
        }

        if (lock) {
            lambdaService.invoke("PUConvocationAttendeeJob")
        }

        distributedLock.release("attendeeLock")

        return Result.Success(
            hashMapOf(
                "code" to ResponseCode.OK,
                "message" to "Attendees List is ${if (lock) "locked" else "unlocked"}."
            )
        )
    }

    suspend fun getAttendees(
        authorizationToken: String?,
        page: Int,
        limit: Int
    ): Result<HashMap<String, Any>, ErrorResponse> {
        if (!authService.isAuthorized(
                role = "read:Attendee",
                principal = authorizationToken
            )
        ) {
            return Result.Error(
                httpStatusCode = HttpStatusCode.Forbidden,
                error = ErrorResponse(
                    errorCode = ResponseCode.NOT_PERMITTED,
                    message = "You don't have privilege to view attendee details."
                )

            )
        }

        val attendees = attendeeRepository.getAttendees(page, limit)

        return Result.Success(
            attendees
        )
    }

    suspend fun searchAttendees(authorizationToken: String?, query: String): Result<List<Attendee>, ErrorResponse> {
        if (!authService.isAuthorized(
                role = "read:Attendee",
                principal = authorizationToken
            )
        ) {
            return Result.Error(
                httpStatusCode = HttpStatusCode.Forbidden,
                error = ErrorResponse(
                    errorCode = ResponseCode.NOT_PERMITTED,
                    message = "You don't have privilege to view attendee details."
                )

            )
        }

        val matchedAttendees = attendeeRepository.searchAttendees(query)

        return Result.Success(matchedAttendees)
    }

}